merkletree/proof.go

```diff
@@ -16,6 +23,75 @@ type ProofNode interface {
 	Commitment() []byte
 }
 
+func computeLeafHash(ap *AuthenticationPath,
+	leafIndex, leafCommitment []byte, leafLevel int, isLeafEmpty bool) (leafHash []byte) {
+	if isLeafEmpty {
+		// empty leaf node
+		leafHash = crypto.Digest(
+			[]byte{EmptyBranchIdentifier},      // K_empty
+			[]byte(ap.TreeNonce),               // K_n
+			[]byte(leafIndex),                  // i
+			[]byte(util.IntToBytes(leafLevel)), // l
+		)
+	} else {
+		// user leaf node
+		leafHash = crypto.Digest(
+			[]byte{LeafIdentifier},             // K_leaf
+			[]byte(ap.TreeNonce),               // K_n
+			[]byte(leafIndex),                  // i
+			[]byte(util.IntToBytes(leafLevel)), // l
+			[]byte(leafCommitment),             // commit(key|| value)
+		)
+	}
+	return
+}
+
+func authPathHash(ap *AuthenticationPath,
+	leafIndex, leafCommitment []byte, leafLevel int, isLeafEmpty bool) []byte {
+	hash := computeLeafHash(ap, leafIndex, leafCommitment, leafLevel, isLeafEmpty)
+	depth := leafLevel - 1
+	indexBits := util.ToBits(leafIndex)
+	for depth > -1 {
+		if indexBits[depth] { // right child
+			hash = crypto.Digest(ap.PrunedTree[depth], hash)
+		} else {
+			hash = crypto.Digest(hash, ap.PrunedTree[depth])
+		}
+		depth -= 1
+	}
+	return hash
+}
+
+// Should verify the vrf index first
+func VerifyAuthPath(ap *AuthenticationPath,
+	leafIndex, leafCommitment []byte, leafLevel int, isLeafEmpty bool,
+	treeHash []byte) bool {
+	// step 1. Verify the auth path of the returned leaf
```

Comment:
Thanks! I thought there were some problems with our change of the order. But maybe I was wrong, so I fixed the order as discussed. Correct me if there still has problems.